DBAgenda.php

```diff
@@ -0,0 +1,311 @@
+<?php
+
+use Monolog\Logger;
+use Sabre\VObject;
+
+abstract class DBAgenda extends Agenda {
+    protected $pdo;
+    
+    public function __construct(string $url, string $username, string $password, object $api, array $agenda_args) {
+        setLogHandlers($this->log);
+        
+        $this->caldav_client = new CalDAVClient($url, $username, $password);
+        $this->api = $api;
+
+        $this->openDB($agenda_args);
+        
+        $this->pdo->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+        $this->pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION); // ERRMODE_WARNING | ERRMODE_EXCEPTION | ERRMODE_SILENT
+
+    }
+    
+    abstract public function createDB();
+    abstract protected function openDB(array $agenda_args);
+    abstract protected function getLastInsertedRowId();
+    
+    public function getUserEventsFiltered(string $userid, array $filters_to_apply = array()) {
+        $sql = 'SELECT vCalendarFilename, number_volunteers_required, vCalendarRaw FROM events WHERE ';
+        $sql .= 'Date(datetime_begin) > :datetime_begin ';
+        
+        $intersect = array();
+        if(isset($filter["my_events"])) {
+            $intersect[] = "SELECT vCalendarFilename FROM events_attendees
+INNER JOIN attendees
+WHERE attendees.email = events_attendees.email and attendees.userid = '$userid'";
+            unset($filter["my_events"]);
+        }
+        
+        if(isset($filter["need_volunteers"])) {
+            $sql .= " number_volunteers_required is not NULL AND ";
```

Comment:
Shouldn't the `AND` be at the start of the line? It seems to that otherwise we would have something like `SELECT ... WHERE Date(....) > ::datetime_being number_volunteers_required is not NULL AND`.

Also, I'm not sure I understand why we only check if this value is not null: shouldn't we also check against the number of attendees already registered?

Reply:
Ok for the misplaced `AND`. Moreover, there were also several other errors in my code. See 5af08b6.

By convention, `number_volunteers_required` is set to `NULL` when the number of required volunteers has not been specified (see `updateEvent()`).
So here, we select events that have a minimum number, of volunteers required, defined .